### PR TITLE
[WIP] Add dropout parameter

### PR DESCRIPTION
**Pull request by carlogrisetti**
I am experimenting with adding a parameter to allow a parametric dropout layer to be included in the "base" model.
Still need to check linting and add the catalog edit to the docs (if present, I don't recall).
Still need to implement that also in tensorflow (it's only torch-side for now, since I'm using that)

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by carlogrisetti** [comment]
@sven1977 @gjoliver would you guys be up for this PR? (tagging you two at random, please forgive me if you're not the correct ones to address)
I reckon dropout is not frequently used in RL, but it might be beneficial in my training, and it's defaulted to none, so... :)

If this could be OK I'll work on adding the same thing in tensorflow

**Comment by gjoliver** [comment]
Hi Carlo,

Thanks for your contribution to RLlib :)

About this particular feature, I am wondering why don't you just override and provide a custom model,
https://docs.ray.io/en/latest/rllib-models.html#custom-models-implementing-your-own-forward-logic
Is this API too complicated for you or are there some other reasons you want the customization to live in the framework?

Thanks again.

**Comment by carlogrisetti** [comment]
I just didn't even think about that, and wanted to stay as "vanilla" as possible, and seeing it's just a few lines I thought cost/benefit was in favor of extending it for everyone (even if almost no one needs it) rather than implementing my own model.

I'll look into the APIs and let you know how I feel about those!

**Comment by gjoliver** [comment]
> I just didn't even think about that, and wanted to stay as "vanilla" as possible, and seeing it's just a few lines I thought cost/benefit was in favor of extending it for everyone (even if almost no one needs it) rather than implementing my own model.
> 
> I'll look into the APIs and let you know how I feel about those!

No problem at all.
I just feel like It may be a bit easier for you if you have ownership of the customization and don't have to wait for our review.

You can see an example of custom model usage here:
https://github.com/ray-project/ray/blob/master/rllib/examples/custom_model_api.py
the actual model is defined in this file:
https://github.com/ray-project/ray/blob/master/rllib/examples/models/custom_model_api.py

**Comment by carlogrisetti** [comment]
It's not what it's easier in my case (it's easier to leave it like this in my branch hehe), I am thinking about what's "right" in terms of a product, and ultimately I've come to the conclusion that this could be a standard parameter you might want to support.

It's not like I am defining a whole different model (like a resnet or some sort of convolutional net), but I see the dropout as an additional "knob" one could use to further tweak a training.
I also see no downside in implementing it in the modelv2, but I might be missing something.
Also the "wait" for it to be integrated is no issue. I am already using the PR code, and do not plan to update Ray in the near future.

That said, if you decide on not implementing that, I will of course proceed and define my custom model.

Thanks!

PS: as a train of thought, if other similar "knobs" appear in the future, I will be more than happy to have those available in the catalog (and not to use those, or maybe just try those in a test training), than to find them out myself (almost impossible to happen) and implement those in a custom model.

**Comment by gjoliver** [comment]
oh completely understand. really appreciate the effort for sure. I can clarify a little bit.

this PR introduce a single layer of dropout, I think, at the end of the fc network? but practically, you can have multiple layers of dropouts throughout the network with different dropout rates.
I can't think of a good way to specify such a generic configuration, similar to how hidden layers are specified for example.

so was thinking this type of experimentation may be easier to configure and maintain using custom model.
Sven may have different ideas, but personally I really hope folks won't have to wait for reviews or branch rllib just to try out a different network shape or configuration.
and model customization, as an officially supported method, seems to be the best approach to achieve that without having to worry about integrating with newer versions, etc.

**Comment by carlogrisetti** [comment]
You are correct indeed... I didn't think about the "dinamicity" of the layer quantity and positioning.

Standing down :)

**Comment by gjoliver** [comment]
really appreciate the effort! :)

**Comment by carlogrisetti** [comment]
I implemented the dropout via a custom model (which basically was a clone of the PR code).
The main feeling was "hell, whatever fix they do on the base code, I won't get it, and will have to periodically check".
Apart from that, it's all OK :)

**Comment by gjoliver** [comment]
How much framework code do you have to duplicate? :)
Ideally, we should allow users to simply override the construction of a torch model object, and nothing else.

**Comment by carlogrisetti** [comment]
Oh well I had to copy and paste rrlib\models\torch\misc.py SlimFC adding the dropout parameter
That one gets used in rrlib\models\torch\misc.py FullyConnectedNetwork adding the parameter pull from the config and adding the parameter to each SlimFC call.
It's not much hehe... I must have gotten too used to Ludwig yaml parametrization of models :)

**Comment by gjoliver** [comment]
haha I see I see. now that you own the entire model logic, you can just hardcode the modifications without parameters :)
